refactoring generate.go queries

diaabadr opened the pull request:
### Description

Enhanced query performance in the `generate.go` by refactoring the insertion process. Replaced individual insertions within a loop with a single insert statement.

### Changes

- Refactored all insertion and update queries to be in single SQL statement
- Added function to utils to extract the values of each statement.

### Related Issues

#358

### Checklist

- [X] Tests included
- [X] Build pass
- [X] Documentation
- [X] Code format and docstring

### Execution time comparison based on the provided constant values in the file:
- **Before:** 84.931081115 seconds
- **After:** 0.993061923 seconds

diaabadr committed:
refactoring generate.go queries

xmonader commented on grid-proxy/tools/db/utils.go:
```diff
@@ -44,6 +45,15 @@ func max(a, b uint64) uint64 {
 	return b
 }
 
+func extractValuesFromInsertQuery(q string) (string, error) {
```
Add test for this function and include docs for the behavior

xmonader commented on grid-proxy/tools/db/generate.go:
```diff
@@ -70,14 +83,32 @@ func generateTwins(db *sql.DB) error {
+
+		value, err := extractValuesFromInsertQuery(insertQuery(&twin))
+		if err != nil {
+			panic(err)
```
the function returns the error, and the caller in generateData is acting on that value, so don't panic here, please refactor to return and the call can do whatever.

xmonader commented on grid-proxy/tools/db/generate.go:
```diff
@@ -361,34 +534,82 @@ func generateNodes(db *sql.DB) error {
+		query := "INSERT INTO location (id, longitude, latitude) VALUES"
+		query += strings.Join(locationValues, ",") + ";"
+		if _, err := db.Exec(query); err != nil {
+			panic(err)
```
same comments as in generateTwins, please apply everywhere.

diaabadr committed:
removing panics

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -70,14 +83,32 @@ func generateTwins(db *sql.DB) error {
 		}
-		if _, err := db.Exec(insertQuery(&twin)); err != nil {
+
+		value, err := extractValuesFromInsertQuery(insertQuery(&twin))
```
i see that insertQuery is only called in extractValuesFromInsertQuery, hence i think it should be called internally if it's needed at all.

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -70,14 +83,32 @@ func generateTwins(db *sql.DB) error {
+		values = append(values, value)
+	}
+
+	if len(values) != 0 {
```
for better readability, wrap this block in a function whose only purpose is to insert twin records

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -91,17 +122,39 @@ func generatePublicIPs(db *sql.DB) error {
 }
 
 func generateFarms(db *sql.DB) error {
+	var values []string
```
rename to something more descriptive

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -117,14 +170,32 @@ func generateFarms(db *sql.DB) error {
 	return nil
 }
 
 func generateContracts(db *sql.DB) error {
```
generate contracts should generate all contracts (node, name, rent). maybe embed the three contract generation functions in it.

mariobassem reviewed (changes requested):
great work Diaa. just some small comments

diaabadr committed:
adding insert function for generate functions
refactor random function
renaming some vars

diaabadr committed:
adding splitted update functions
adding tests for utils
documenting util functions

diaabadr committed:
modifying rand function
adding test for rnd function

mariobassem commented on grid-proxy/tools/db/types.go:
```diff
@@ -54,6 +54,8 @@ type twin struct {
 	relay        string
 	public_key   string
 }
+
```
remove new line

mariobassem commented on grid-proxy/tools/db/utils.go:
```diff
@@ -2,21 +2,32 @@ package main
 
-func rnd(min, max uint64) uint64 {
-	return rand.Uint64()%(max-min+1) + min
+var errMaxMinInvalid = errors.New("min cannot be greater than max")
```
we don't really need to create an error variable, right?
if it's to easily check it when testing, just checking if there is an error is usually enough.

mariobassem commented on grid-proxy/tools/db/utils_test.go:
```diff
@@ -0,0 +1,39 @@
+	assert.Equal(t, want, got)
+}
+
+func TestObjectToTupleString(t *testing.T) {
```
great!
maybe just add one more test run in this function to test NULL values?

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -31,10 +31,10 @@ const (
 	nodeCount            = 1000
 	farmCount            = 100
-	normalUsers          = 2000
+	normalUsers          = 200
```
why did normal users count decrease to 200??

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -31,10 +31,10 @@ const (
 	nodeUpRatio          = .5
 	nodeCount            = 1000
```
maybe increase the numbers a bit??
let's do nodes 100000, farms 10000, nodeContracts 100000, rentContracts 1000, nameContracts 1000, normalUsers 50000. 
what do you think?

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -397,37 +606,113 @@ func generateNodeGPUs(db *sql.DB) error {
+func generateContracts(db *sql.DB) error {
+	billCount := 1
+	// Note: it's order senstive
+	if err := generateRentContracts(db, &billCount); err != nil {
```
having to pass the billCount by reference works in our use case, but in general it's not a really good idea unless it's needed, as a pointer creates the possibility for the value to not even be there.
i believe we can work around that, right?

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -397,37 +606,113 @@ func generateNodeGPUs(db *sql.DB) error {
+
+func generateContracts(db *sql.DB) error {
+	billCount := 1
+	// Note: it's order senstive
```
is there a way to make it not order sensitive?

diaabadr committed:
increasing counts
adding test with null values for extractTuple function
making contracts generation not order senstive

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -70,38 +82,66 @@ func generateTwins(db *sql.DB) error {
 
 func generatePublicIPs(db *sql.DB) error {
+	var publicIPs []string
+	var nodeContracts []interface{}
```
does nodeContracts need to be []interface{} ?

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -70,38 +82,66 @@ func generateTwins(db *sql.DB) error {
 		}
 		ip := randomIPv4()
+		farmID, err := rnd(1, farmCount)
+
```
remove empty line?

diaabadr replied:
formatter does not remove them.

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -216,47 +324,74 @@ func generateNameContracts(db *sql.DB) error {
+	return nil, billsStartID, contractsStartID
 }
-func generateRentContracts(db *sql.DB) error {
+func generateRentContracts(db *sql.DB, billsStartID, contractsStartID int) (error, int, int) {
```
nothing mandates that an error should have a specific position an a function's returned values, but if we care about our code's readability we would stick with the convention of having the error as the last returned value, right?

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -397,37 +609,117 @@ func generateNodeGPUs(db *sql.DB) error {
+	billsStartID := 1
+	contractsStartID := 1
+
+	err, billsStartID, contractCount := generateRentContracts(db, billsStartID, contractsStartID)
```
this is better from last version. 
i would go further with this and make the generateRentContract, generateNodeContracts, and generateNameContracts functions just create the string arrays of the contracts and the bill reports, and do the actual insert calls here in the generateContracts function.
but this is just nitpicking, feel free to ignore.

diaabadr replied:
what about generating all contracts in the functions and collecting reports from each one and inserting them once? that's the one which will make the performance better as inserting all bills at once.

diaabadr committed:
removing interface not necessary
returning error as last attr returned

mariobassem commented on grid-proxy/tools/db/generate.go:
```diff
@@ -267,72 +399,131 @@ func generateRentContracts(db *sql.DB) error {
+		}
+		updatedAt := time.Now().Unix() - int64(periodFromLatestUpdate)
+
+		periodFromLatestUpdate, err = rnd(0, 60*40*1)
```
should this be in the `if up {}` block ??

diaabadr committed:
adding lastupdate to up block

diaabadr committed:
making insertion of bill reports together

diaabadr committed:
removing startBillID

diaabadr committed:
logging errors

diaabadr committed:
adding old numbers

diaabadr committed:
removing panic from objectToTuple util function
